source/server/hot_restart_impl.cc

```diff
@@ -91,10 +90,20 @@ void initializeMutex(pthread_mutex_t& mutex) {
   pthread_mutex_init(&mutex, &attribute);
 }
 
-HotRestartImpl::HotRestartImpl(const Options& options)
-    : as_child_(HotRestartingChild(options.baseId(), options.restartEpoch())),
-      as_parent_(HotRestartingParent(options.baseId(), options.restartEpoch())),
-      shmem_(attachSharedMemory(options)), log_lock_(shmem_->log_lock_),
+// N.B. Older versions of Envoy would immediately multiply the base id from the
+// command line by 10 in OptionsImpl. Very old versions of Envoy used this
+// convention to provide space between consecutive ids for the parent & child
+// domain socket names. Currently, Envoy uses the restart epoch to distinguish
+// a sequence of hot-restart parent/child domain socket names and thus the
+// scaling is not necessary. It remains for backwards compatibility with
+// versions that perform the scaling.
+// TODO(zuercher): Figure out if there's a way to undo this scaling mess with having to
+// bump the hot restart version.
```

Comment:
We actually still depend on this.

As history: We used to deal with 3 sockets addresses formatted as `envoy_domain_socket_{N}` where N was `X = (base_id*10 + restart_epoch)`, `X-1` or `X+1`. X-1 was the parent's socket (only used if restart epoch != 0), X was the socket we listened on, X+1 was the future child's address.

In 1.11.0 we switched to use 4 sockets addresses formatted as `envoy_domain_socket_{ROLE}_{N}`. We send to the parent process on `(parent, X-1)` (again only if restart epoch != 0), receive responses from that parent on `(child, X)`. We listen on `(parent, X)` for future hot restart messages and respond to them via `(child, X+1)`.

That's all fine and works well enough for a single base id. The problem is that for a given base_id, B, after 10 hot restarts, the value of X (as described above) collides with an existing Envoy that had a base id B+1, but hasn't been hot restarted. 

I think it manifests as the hot-restarting Envoy failing to open the `(parent, X)` socket and presumably the restart fails and falls back to staring a new process, which resets the restart epoch to 0. Perhaps worse is that if you wait to start the second Envoy till the first one has hot restarted 10 times, the second one won't be able to start because it can't open the socket.

What I'd like to do is make the socket name `envoy_domain_socket_{ROLE}_{BASE_ID}_{X}` and just have X be the restart_epoch (or +1 or -1, depending on which address we need). That version needn't multiple the base id by 10. I think that's another PR, though.

Reply:
I updated the comment. when I wrote it, I had somehow convinced myself that we no longer were able to collide, but that's not correct. We just made unambiguous the socket addresses used for communication between the current/previous process vs. those used for the current/future process.

Reply:
Ah OK, thanks for the detailed explanation. It might be possible to change ^ without bumping the compat version by trying both types of addresses for an intermediate amount of time? Agreed on a different PR though.